proposals/254-padding-negotiation.txt

```diff
@@ -71,331 +77,66 @@ the circuit.
-sends to the upstream relay.
-
-On the client, the "SENT" direction means packets destined towards the
-relay, where as "RECV" means packets destined towards the client.
-However, on the relay, the "SENT" direction means packets destined
-towards the client, where as "RECV" means packets destined towards the
-relay.
-
-3.2.2. The original Adaptive Padding algorithm
-
-As we have noted, the state machines above represent a generalization of
-the original Adaptive Padding algorithm. To implement the original
-behavior, the following flags should be set in both the client and
-the relay state machines:
-
- num_machines = 1;
-
- machines[0].transition_burst_events =
-    RELAY_PADDING_TRANSITION_EVENT_NONPADDING_SENT;
-
- machines[0].burst.transition_reschedule_events =
-    RELAY_PADDING_TRANSITION_EVENT_NONPADDING_SENT;
-
- machines[0].burst.transition_gap_events =
-    RELAY_PADDING_TRANSITION_EVENT_PADDING_SENT;
+Each state of a padding machine can react to the following cell events:
+  * Non-padding cell received
+  * Padding cell received
+  * Non-padding cell sent
+  * Padding cell sent
 
- machines[0].burst.transition_start_events =
-    RELAY_PADDING_TRANSITION_EVENT_INFINITY;
+Additionally, padding machines emit the following internal events to themselves:
+  * Infinity bin was selected
+  * The histogram bins are empty
+  * The length count for this state was exceeded
 
- machines[0].gap.transition_reschedule_events =
-    RELAY_PADDING_TRANSITION_EVENT_PADDING_SENT;
+Each state of the padding machine specifies a set of these events that cause
+it to cancel any pending padding, and a set of events that cause it to
+transition to another state, or transition back itself.
 
- machines[0].gap.transition_burst_events =
-    RELAY_PADDING_TRANSITION_EVENT_NONPADDING_SENT |
-    RELAY_PADDING_TRANSITION_EVENT_INFINITY;
+When an event causes a transition to a state (or back to the same state), a
+delay is sampled from the histogram or delay distribution, and padding cell is
+scheduled to be sent after that delay.
 
-The rest of the transition fields would be 0.
+If a non-padding cell is sent before the timer, the timer is cancelled and a
+new padding delay is chosen.
 
-Adding additional transition flags will either increase or decrease the
-amount of padding sent, depending on their placement.
+3.1.1. Histogram Specification
 
-The second machine slot is provided in the event that it proves useful
-to have separate state machines reacting to both sent and received
-traffic.
-
-3.2.3. Histogram decoding/representation
-
-Each of the histograms' fields represent a probability distribution that
-is expanded into bins representing time periods a[i]..b[i] as follows:
+If a histogram is used by a state (as opposed to a fixed parameterized
+distribution), then each of the histograms' fields represent a probability
+distribution that is expanded into bins representing time periods a[i]..b[i]
+as follows:
 
```

Comment:
I think this approach of using pseudocode to specify how the histograms work is not very helpful. Ideally we should be able to explain the concept with words.

Reply:
Converted into prose in https://github.com/torproject/torspec/pull/41/commits/1dd7f1ff78fcb69121130cbd802b0b8b527ffc63